### PR TITLE
Eng 10842 mixed sph

### PR DESCRIPTION
**Pull request by luyangco**
This patch provides VoltDB the ability to start with mixed sitesperhost settings. By using the hidden option `-s` or `--sitesperhost` the user can overrides the `sitesperhost` field in deployment file for each host.

The Pro-side pull request is https://github.com/VoltDB/pro/pull/1534

### CODE REVIEW

File: src/frontend/org/voltcore/messaging/HostMessenger.java [hunk 1]
```diff
@@ -90,7 +94,7 @@
     private static final VoltLogger m_hostLog = new VoltLogger("HOST");
     private static final VoltLogger m_tmLog = new VoltLogger("TM");
 
-    public static final CopyOnWriteArraySet<Long> VERBOTEN_THREADS = new CopyOnWriteArraySet<Long>();
+    public static final CopyOnWriteArraySet<Long> VERBOTEN_THREADS = new CopyOnWriteArraySet<>();
```
Comment (luyangco) [inline src/frontend/org/voltcore/messaging/HostMessenger.java]: Sorry for the inconvenience, I'll change my Eclipse setting for future peer review, however I don't know to get it back in this pull request.

File: src/frontend/org/voltdb/compiler/VoltCompiler.java [hunk 2]
```diff
@@ -676,7 +676,13 @@ private InMemoryJarfile compileInternal(
             Deployment deployment = catalogContext != null ? catalogContext.cluster.getDeployment().get("deployment") : null;
             int hostcount = clusterSettings != null ? clusterSettings.hostcount() : 1;
             int kfactor = deployment != null ? deployment.getKfactor() : 0;
-            int sitesPerHost = deployment != null ? deployment.getSitesperhost() : 8;
+            int localSitesCount;
```
Comment (luyangco) [inline src/frontend/org/voltdb/compiler/VoltCompiler.java]: I'll do it.

File: lib/python/voltcli/voltdb.d/start.py [hunk 3]
```diff
@@ -37,6 +37,7 @@
         VOLT.StringOption('-D', '--dir', 'directory_spec', voltdbroot_help, default = None),
         VOLT.BooleanOption('-r', '--replica', 'replica', 'start replica cluster', default = False),
         VOLT.BooleanOption('-A', '--add', 'enableadd', 'allows the server to elastically expand the cluster if the cluster is already complete', default = False),
+        VOLT.IntegerOption('-s', '--sitesperhost', 'sitesperhost', ''),
```
Comment (luyangco) [inline lib/python/voltcli/voltdb.d/start.py]: Good catch! "-s" IS in help output, I'll remove it.

File: src/frontend/org/voltdb/VoltZK.java [hunk 4]
```diff
@@ -56,6 +56,7 @@
     public static final String operationMode = "/db/operation_mode";
     public static final String exportGenerations = "/db/export_generations";
     public static final String importerBase = "/db/import";
+    public static final String sitesPerHost = "/db/sitesperhost";
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/VoltZK.java]: I see a trend that we are crazy adding more and more metadata on ZK.
Better to add comments explain what this data is for.

File: src/frontend/org/voltdb/SnapshotSaveAPI.java [hunk 5]
```diff
@@ -115,7 +115,10 @@ public VoltTable startSnapshotting(
     {
         TRACE_LOG.trace("Creating snapshot target and handing to EEs");
         final VoltTable result = SnapshotUtil.constructNodeResultsTable();
-        final int numLocalSites = context.getCluster().getDeployment().get("deployment").getSitesperhost();
+        final int numLocalSites = VoltDB.instance().getHostMessenger().getLocalSitesCount();
+        if (numLocalSites == -1) {
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/SnapshotSaveAPI.java]: Now It's better to use a FINAL CONSTANT value here. We use -1 multiple times around.

File: src/frontend/org/voltdb/iv2/LeaderAppointer.java [hunk 6]
```diff
@@ -163,6 +163,7 @@ public void run(List<String> children)
                         int pid = aPartition.getInt("partition_id");
                         if (pid == m_partitionId) {
                             replicaCount = aPartition.getJSONArray("replicas").length();
+                            break;
```
Comment (xinjiacs) [inline src/frontend/org/voltdb/iv2/LeaderAppointer.java]: This file contains all style changes other than the early break from FOR loop. 

It's good to break the for loop earlier. Better to be in the separate pull request.

File: src/frontend/org/voltdb/compiler/ClusterConfig.java [hunk 7]
```diff
@@ -29,19 +29,20 @@
...
 import org.json_voltpatches.JSONStringer;
 import org.voltcore.logging.VoltLogger;
 import org.voltdb.VoltDB;
+import org.voltdb.utils.MiscUtils;
 
+import com.google_voltpatches.common.collect.Lists;
+import com.google_voltpatches.common.collect.Maps;
 import com.google_voltpatches.common.collect.Multimap;
-import org.voltdb.utils.MiscUtils;
+import com.google_voltpatches.common.collect.Sets;
```
Comment (luyangco) [inline src/frontend/org/voltdb/compiler/ClusterConfig.java]: Sure I can do that, actually the Chapter 7 of https://github.com/VoltDB/voltdb/wiki/Setting-up-an-Eclipse-Project recommends people to do it.

File: src/frontend/org/voltdb/compiler/ClusterConfig.java [hunk 8]
```diff
@@ -124,10 +125,31 @@ public static void addPartitions(JSONObject topo, Multimap<Integer, Integer> par
...
 
-    public ClusterConfig(int hostCount, int sitesPerHost, int replicationFactor)
+    /**
+     * Add new sitesperhost mapping to the toplogy
+     * @param topo              The topology that will be added to.
+     * @param joinedHostIds     Host ids that are elastically joining into cluster.
+     * @param sphMap            A map of host ids to their sitesperhost setting
+     * @throws JSONException
+     */
+    public static void addSitesPerHosts(JSONObject topo,
```
Comment (luyangco) [inline src/frontend/org/voltdb/compiler/ClusterConfig.java]: It is used in  `ElasticJoinCoordinator.java`  `activateJoin()`

File: src/frontend/org/voltdb/CatalogContext.java [hunk 9]
```diff
@@ -314,14 +315,19 @@ public DeploymentType getDeployment()
...
-        int sitesPerHost = deployment.getSitesperhost();
+        Map<Integer, Integer> sphMap = VoltDB.instance().getHostMessenger().getSitesPerHostFromZK();
+        int totalSitesCount = 0;
+        for (Map.Entry<Integer, Integer> e : sphMap.entrySet()) {
+            totalSitesCount += e.getValue();
+        }
+        int localSitesCount = sphMap.get(VoltDB.instance().getHostMessenger().getHostId());
         int kFactor = deployment.getKfactor();
         logLines.put("deployment1",
                 String.format("Cluster has %d hosts with leader hostname: \"%s\". %d sites per host. K = %d.",
```
Comment (luyangco) [inline src/frontend/org/voltdb/CatalogContext.java]: Will do

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 10]
```diff
@@ -1090,12 +1100,10 @@ public void initialize(Configuration config) {
                 }
                 m_clientInterface = ClientInterface.create(m_messenger, m_catalogContext, m_config.m_replicationRole,
                         m_cartographer,
-                        m_configuredNumberOfPartitions,
                         clientIntf,
                         config.m_port,
                         adminIntf,
-                        config.m_adminPort,
-                        m_config.m_timestampTestingSalt);
+                        config.m_adminPort);
```
Comment (luyangco) [inline src/frontend/org/voltdb/RealVoltDB.java]: Right, those two parameters were unused.

File: src/frontend/org/voltdb/SnapshotSaveAPI.java (hunk 5 above)
Comment (luyangco) [inline src/frontend/org/voltdb/SnapshotSaveAPI.java]: Will do.

File: tests/frontend/org/voltdb/compiler/TestClusterCompiler.java [hunk 11]
```diff
@@ -27,22 +27,27 @@
...
 
 import junit.framework.TestCase;
-import org.voltdb.VoltDB;
 
 public class TestClusterCompiler extends TestCase
 {
     public void testNonZeroReplicationFactor() throws Exception
     {
-        ClusterConfig config = new ClusterConfig(3, 1, 2);
+        Map<Integer, Integer> sphMap = Maps.newHashMap();
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/compiler/TestClusterCompiler.java]: If you create a static method with input parameter: hostCount, SPH
You can refactor many of this copy past map constructing into one line.

File: tests/frontend/org/voltdb/TestMixedSitesPerHostCluster.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        Map<Integer, Integer> subOptimalSphMap = Maps.newHashMap();
+        subOptimalSphMap.put(0, 2);
+        subOptimalSphMap.put(1, 6);
+        subOptimalSphMap.put(2, 3);
+        cluster = new MixedSitesPerHostCluster(subOptimalSphMap);
+        try {
+            cluster.start();
+            fail();
+        }
+        catch (RuntimeException e) {}
```
Comment (xinjiacs) [inline tests/frontend/org/voltdb/TestMixedSitesPerHostCluster.java]: It's better to verify the error message as well. Then we know why it fails. Not other reasons that not expected.
